### PR TITLE
SZ sugar

### PR DESCRIPTION
**Pull request by damonge**
Adds capabilities to compute tSZ-related quantities.

Things left to do:

- [x] Fourier transform FFTLog.
- [x] GNFW instead of just Planck2013.
- [x] Check documentation.
- [x] Finish unit test with class_sz.

### CODE REVIEW

**Comment by tilmantroester** [comment]
If you need benchmark code, I got some for the tSZ kernel.

**Comment by damonge** [comment]
Ah great. You were already on my list of suspects for this!

**Comment by damonge** [comment]
OK @tilmantroester what kind of benchmarks could you provide?
Ideally something like C_ell^yy and C_ell^ym would be great... Note that I've only implemented the Arnaud GNFW profile here for now

**Comment by tilmantroester** [comment]
What I could provide easily is a the tSZ redshift kernel. Cls would require a bit more work. I have 
 C_ell^ym in principle from CosmoSIS but those would be based on a non-UPP halo model (Alex Mead's HMx). The python interface is not yet stable though. But in principle that could be plugged into the generalised tracers in CCL to get C_ell^ym to compare.

**Comment by damonge** [comment]
@tilmantroester would you mind reviewing this?

I've benchmarked the full power spectrum calculation against [class_sz](https://github.com/borisbolliet/class_sz), after iterating with @borisbolliet (who has been most helpful!).

@beckermr I've introduced functionality to update the internal parameters of halo profiles through [this kind of stuff](https://github.com/LSSTDESC/CCL/blob/bfcf3aa14ba0f30d180500a98f288a2faad41951/pyccl/halos/profiles.py#L922). Having will be very useful, since many of these parameters are "fast" and shouldn't trigger a recalculation of slow things like Fourier transforms (this will be also true when we implement HODs and other profiles). I know you don't like kwargs, so I'd like to hear your thoughts on this particular point (no need to look at the full PR).

**Comment by beckermr** [comment]
@damonge Why do you need the kwargs? Just list each one and set it to None. Then you can test and if it is non-None, then reset it.

**Comment by damonge** [comment]
yep, that would work too. I just wondered if kwargs would be useful for likelihood design (e.g. if all the likelihood parameters are in a dictionary, you can just pass the full dictionary through, without having to pick the params in it that this particular profile depends on)

**Comment by beckermr** [comment]
This sounds convenient but ultimately it is a recipe for errors. By making things explicit, we can help catch cases where ppl pass things that they think do one thing but don't actually do anything at all.

**Comment by damonge** [comment]
Alright, let me do this.

**Comment by damonge** [comment]
Alright, done. @tilmantroester , let me know if you think you'll be able to take a look.

File: pyccl/tests/test_profiles.py [hunk 1]
```diff
@@ -70,6 +70,21 @@ def test_empirical_smoke(prof_class):
...
+    smoke_assert_prof_real(p)
+    p.update_parameters(mass_bias=0.7,
+                        alpha=1.24)
+    assert p.alpha == 1.24
+    assert p.mass_bias == 0.7
+    assert p.beta == beta_old
+    for n in ['P0', 'P0_hexp', 'alpha',
+              'beta', 'gamma', 'alpha_P',
+              'c500', 'mass_bias']:
+        p.update_parameters(**{n: 1.})
```
Comment (tilmantroester) [inline pyccl/tests/test_profiles.py:85]: Does this just test if `update_parameters` accepts `n` as a kwarg?

File: pyccl/tracers.py [hunk 2]
```diff
@@ -542,6 +543,38 @@ def __init__(self, cosmo, z_source, n_samples=100):
...
+        n_chi (float): number of intervals in the radial comoving
+            distance on which we sample the kernel.
+    """
+    def __init__(self, cosmo, z_max=6., n_chi=1024):
+        self.chi_max = comoving_radial_distance(cosmo, 1./(1+z_max))
+        chi_arr = np.linspace(0, self.chi_max, n_chi)
+        a_arr = scale_factor_of_chi(cosmo, chi_arr)
+        # This is \sigma_T / (m_e * c^2)
+        prefac = 4.01710079e-06
+        w_arr = prefac * a_arr
```
Comment (tilmantroester) [inline pyccl/tracers.py:572]: The kernel depends on whether the pressure is in physical or comoving coordinates. The docstrings don't specify, is that documented elsewhere?
This has tripped my up before when trying to match factors of `a` to hydrosim quantities.

File: pyccl/tests/test_profiles.py (hunk 1 above)
Comment (damonge) [reply pyccl/tests/test_profiles.py]: It did, but that's not great, so now it also tests that the parameter actually gets updated (which in turn has allowed me to fix a minibug).

File: pyccl/tracers.py (hunk 2 above)
Comment (damonge) [reply pyccl/tracers.py:573]: Right, this should be physical, I believe (as in the one that goes in Eq. 2 of https://arxiv.org/pdf/1909.09102.pdf). I've documented this.

**Comment by damonge** [comment]
Thanks @tilmantroester ! Back to you.

File: pyccl/halos/profiles.py [hunk 3]
```diff
@@ -843,3 +844,214 @@ def _real(self, cosmo, r, M, a, mass_def):
...
+    mass :math:`M_{500}` via the mass bias parameter :math:`(1-b)`
+    as :math:`\\tilde{M}_{500}=(1-b)M_{500}`. :math:`\\tilde{r}_{500}`
+    is the overdensity halo radius associated with :math:`\\tilde{M}_{500}`
+    (note the intentional tilde!), and the profile is defined for
+    a halo overdensity :math:`\\Delta=500` with respect to the
+    critical density.
+
+    The default arguments (other than `mass_bias`), correspond to the
+    profile parameters used in the Planck 2013 (V) paper.
+
```
Comment (tilmantroester) [inline pyccl/halos/profiles.py:879]: Could you add the units of the pressure profile here as well for completeness?

Comment (damonge) [reply pyccl/halos/profiles.py:880]: Good point. Done!

**Comment by damonge** [comment]
Awesome, thanks!

**Pull request merged**